test/mongoose_subdomain_core_SUITE.erl

```diff
@@ -0,0 +1,567 @@
+    Domains = [<<"subdomain.example.net">>, <<"subdomain.example.org">>,
+               <<"example.net">>, <<"example.org">>],
+    [mongoose_domain_core:delete(Domain) || Domain <- Domains],
+    %%------------------------------------------
+    %% testing type #2 subdomain names collision
+    %%------------------------------------------
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                                Pattern2, Handler)),
+    mongoose_domain_core:insert(<<"some.fqdn">>, ?DYNAMIC_HOST_TYPE2, dummy_src),
+    mongoose_domain_core:insert(<<"another.fqdn">>, ?DYNAMIC_HOST_TYPE2, dummy_src),
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                                Pattern3, Handler)),
+    ?assertEqualLists([<<"some.fqdn">>, <<"another.fqdn">>], get_all_subdomains()),
+    ?assertEqualLists(
+        [<<"some.fqdn">>, <<"another.fqdn">> | ?DYNAMIC_DOMAINS],
+        mongoose_domain_core:get_domains_by_host_type(?DYNAMIC_HOST_TYPE2)),
+    no_subdomain_collisions(),
+    {ok, SubdomainInfo2} =
+        mongoose_subdomain_core:get_subdomain_info(<<"another.fqdn">>),
+    ?assertEqual(
+        [{check_domain_name, [?DYNAMIC_HOST_TYPE2, <<"some.fqdn">>]},
+         {check_subdomain_name, [SubdomainInfo2]}],
+        get_list_of_domain_collisions()).
+
+%%-------------------------------------------------------------------
+%% internal functions
+%%-------------------------------------------------------------------
+get_subdomains_table() ->
+    ets:tab2list(mongoose_subdomain_core).
+
+get_subdomains_table_size() ->
+    ets:info(mongoose_subdomain_core, size).
+
+get_all_subdomains() ->
+    %% mongoose_subdomain_core table is indexed by subdomain name field
+    KeyPos = ets:info(mongoose_subdomain_core, keypos),
+    [element(KeyPos, Item) || Item <- ets:tab2list(mongoose_subdomain_core)].
+
+make_wrapper_fn(N, M) when N > M ->
+    %% the wrapper function generates a new loop processing function
+    %% that pauses after after processing N domains, removes M of the
+    %% already processed domains and resumes after that.
+    fun(Fn) ->
+        put(number_of_iterations, 0),
+        fun(HostType, DomainName) ->
+            NumberOfIterations = get(number_of_iterations),
+            if
+                NumberOfIterations =:= N -> remove_some_domains_asyn(M);
+                true -> ok
+            end,
+            put(number_of_iterations, NumberOfIterations + 1),
+            Fn(HostType, DomainName)
+        end
+    end.
+
+remove_some_domains_asyn(N) ->
+    {Pid, Ref} = spawn_monitor(fun() -> remove_some_domains(N) end),
+    receive
+        {'DOWN', Ref, process, Pid, _Info} -> ok
+    end.
+
+remove_some_domains(N) ->
+    AllSubdomains = get_all_subdomains(),
+    IndexesToRemove = list_of_rand_unique_ints(N,length(AllSubdomains)),
```

Comment:
Please make this deterministic and predictable, so that a repeated test does not give different results. The main motivation is to make debugging easier.